rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,146 @@
+on. The wrapper will follow all the security and spaces filtering of saved
+objects so that only users with appropriate permissions will be able to obtain
+the scrubbed objects or _save_ objects with encrypted attributes.
+
+## Possible Problems
+
+Consumers of scrubbed objects may try to modify them causing the encrypted
+attributes to be dropped. But this could be prevented by disallowing updates
+from the default saved objects client. And instead force all updates to go
+through the plugin directly.
+
+## Decrypting attributes
+
+In order to get the resulting decrypted attributes back the plugin would require
+a method be registered which would be passed the saved object including the
+decrypted attributes then return an invokable method that would perform the
+retrieval of the saved object the decryption and invoking of the desired method.
+
+```JS
+const actionMethod = server.plugins.encryptedAttributes.buildDecryptionContext(id, type, (savedObject) => {
+  const { username, password } = savedObject.secretsParams;
+  ... // do integration with third party
+});
+... // later
+actionMethod(); // retrieve, decrypt, and invoke above method
+```
+
+## Benefits
+
+No explicit access to a method that takes in an encrypted string exists. And no
+access to the private key used is exposed through the plugin. If the type was not
+registered no decryption is possible. No need to handle the saved object with
+the encrypted attributes reducing the risk of accidentally returning it in a
+handler.
+
+# Drawbacks
+
+- This would add the ability to encrypt data on any saved object in the system.
+- Possibly add performance issues to retrieving saved objects.
+- Might be complicated to test along with spaces and security
+- Could be too complicated in general
+- The attributes that are encrypted have to be defined and if they change they
+  need to be migrated.
+
+# Alternatives
+
+Only allow this to be used within the Actions service itself where the details
+of the saved object are handled there directly. And the saved objects are
+`hidden` but still use the security and spaces wrappers.
+
+# Adoption strategy
+
+Integration should be pretty easy which would include depending on the plugin.
+And registering the desired saved object type with it. And building and using
+the desired decryption contexts.
+
+# How we teach this
+
+`encryptedAttributes` as the name of the `thing` where it's seen as a separate
+extension on top of the saved object service.
+
+Provide a readme.md in the plugin directory with examples for how to depend on
+the plugin and define a type has hidden attributes.
+
+# Unresolved questions
+
+- Is `encryptedAttributes` an acceptable name for the plugin
+- Is building a method that receives the object with the decrypted attributes
+  make sense?
+- Are there other use-cases that are not served with that interface?
+- How would this work with migrations, if the attribute names wanted to be
```

Comment:
I am not sure about the restrictions on whose allowed to define migrations. @mikecote might know more. I would think those plugins that define a saved object type would provide the necessary migration code for the encrypted attributes. My concern is if the migration functions will have access to the decryption method

Reply:
As of 7.1 we're enforcing migrations to be defined where the mappings are defined. The PR implementing this can be found here https://github.com/elastic/kibana/pull/31739.